From: Cristian Marussi <cristian.marussi@arm.com>
Date: Wed, 28 Jun 2023 18:43:34 +0100
Subject: [PATCH] dt-bindings: mailbox: arm,mhuv3: Add bindings

Add bindings for the ARM MHUv3 Mailbox controller, limited to the MHUv3
Doorbell Extension.

Signed-off-by: Cristian Marussi <cristian.marussi@arm.com>

Upstream-Status: Pending
  Dependency on MHUv3 specification being made public.
---
 .../bindings/mailbox/arm,mhuv3.yaml           | 161 ++++++++++++++++++
 1 file changed, 161 insertions(+)
 create mode 100644 Documentation/devicetree/bindings/mailbox/arm,mhuv3.yaml

diff --git a/Documentation/devicetree/bindings/mailbox/arm,mhuv3.yaml b/Documentation/devicetree/bindings/mailbox/arm,mhuv3.yaml
new file mode 100644
index 000000000000..d9cb35403059
--- /dev/null
+++ b/Documentation/devicetree/bindings/mailbox/arm,mhuv3.yaml
@@ -0,0 +1,161 @@
+# SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+%YAML 1.2
+---
+$id: http://devicetree.org/schemas/mailbox/arm,mhuv3.yaml#
+$schema: http://devicetree.org/meta-schemas/core.yaml#
+
+title: ARM MHUv3 Mailbox Controller
+
+maintainers:
+  - Sudeep Holla <sudeep.holla@arm.com>
+  - Cristian Marussi <cristian.marussi@arm.com>
+
+description: |
+  The Arm Message Handling Unit (MHU) Version 3 is a mailbox controller that
+  enables unidirectional communications with remote processors through various
+  possible transport protocols.
+  The controller can optionally support a varying number of extensions that, in
+  turn, enable different kinds of transports to be used for communication.
+  Number, type and characteristics of each supported extension can be discovered
+  dynamically at runtime.
+
+  Given the unidirectional nature of the controller, an MHUv3 mailbox controller
+  is composed of a PostBox (PBX) transmitter and a MailBox (MBX) receiver block.
+  A PBX block can only be written to, while an MBX block can only be read from;
+  both need to be present and defined if you need to establish a bidirectional
+  communication, since you will have to acquire two distinct unidirectional
+  channels, one for each block.
+  As a consequence both blocks needs to be represented separately and specified
+  as distinct nodes in order to properly describe their resources.
+  Note that, though, due to the runtime discoverability, there is no need to
+  identify the type of blocks with distinct compatibles.
+
+  Following are the currently supported extensions.
+
+  - Doorbell (DBE): Each transfer is made up of single bit flag, using any one
+    of the bits in a channel window. A channel window can support up to 32
+    doorbells and the entire window shall be used in doorbell protocol.
+    Optionally, data may be transmitted through an out-of-band shared memory
+    region, wherein the MHU is used strictly as an interrupt generation
+    mechanism but that is out of the scope of these bindings.
+    When DBE is implemented, it allows for a maximum of 128 channels.
+
+properties:
+  compatible:
+    description: |
+      Mailbox controller implementing Arm MHUv3 Postbox or Mailbox block.
+    contains:
+      const: arm,mhuv3
+
+  reg:
+    description: |
+      Base address of the MHUv3 controller PBX or MBX block.
+    maxItems: 1
+
+  interrupts:
+    description: |
+      The MHUv3 controller always implements at least a combined MBX interrupt
+      in the "receiver" mode to signal that a new Transfer event has been
+      initiated from the PBX and a combined PBX interrupt on the "sender" side
+      to let the MBX actively signal that a Transfer has been acknowledged by
+      the MBX.
+    maxItems: 1
+
+  interrupt-names:
+    oneOf:
+      - const: combined-pbx
+        description: PBX Combined interrupt
+      - const: combined-mbx
+        description: MBX Combined interrupt
+
+  clocks:
+    maxItems: 1
+
+  clock-names:
+    maxItems: 1
+
+  '#mbox-cells':
+    description: |
+      It is always set to 3. The first argument in the consumers 'mboxes'
+      property represents the extension type, the second is for the channel
+      window number to use and the third depends on extension type.
+      Extension type for DBE is 0 with the third parameter representing the
+      doorbell number to use.
+
+      mboxes = <&mhu 0 0 5>; // DBE, Doorbell Channel Window 0, doorbell 5.
+      mboxes = <&mhu 0 1 7>; // DBE, Doorbell Channel Window 1, doorbell 7.
+    const: 3
+
+required:
+  - compatible
+  - reg
+  - interrupts
+  - interrupt-names
+  - '#mbox-cells'
+
+additionalProperties: false
+
+examples:
+  - |
+    soc {
+        #address-cells = <2>;
+        #size-cells = <2>;
+
+        mhu_tx: mailbox@2aaa0000 {
+            compatible = "arm,mhuv3";
+            #mbox-cells = <3>;
+            reg = <0 0x2aaa0000 0 0x10000>;
+            clocks = <&clock 0>;
+            clock-names = "apb_pclk";
+            interrupt-names = "combined-pbx";
+            interrupts = <0 36 4>;
+        };
+
+        mhu_rx: mailbox@2ab00000 {
+            compatible = "arm,mhuv3";
+            #mbox-cells = <3>;
+            reg = <0 0x2aab0000 0 0x10000>;
+            clocks = <&clock 0>;
+            clock-names = "apb_pclk";
+            interrupt-names = "combined-mbx";
+            interrupts = <0 35 4>;
+        };
+
+        firmware {
+            scmi {
+                compatible = "arm,scmi";
+                mboxes = <&mhu_tx 0 0 0>, <&mhu_rx 0 0 0>, <&mhu_rx 0 0 1>;
+                shmem = <&cpu_scp_lpri0>, <&cpu_scp_lpri1>;
+
+                #address-cells = <1>;
+                #size-cells = <0>;
+
+                scmi_devpd: protocol@11 {
+                    reg = <0x11>;
+                    #power-domain-cells = <1>;
+                };
+
+                scmi_dvfs: protocol@13 {
+                    reg = <0x13>;
+                    #clock-cells = <1>;
+
+                    mboxes = <&mhu_tx 0 1 0>, <&mhu_rx 0 1 0>, <&mhu_rx 0 1 1>;
+                    shmem = <&cpu_scp_hpri0>, <&cpu_scp_hpri1>;
+                };
+            };
+        };
+
+        mhu_client: dsp@596e8000 {
+            compatible = "fsl,imx8qxp-dsp";
+            reg = <0 0x596e8000 0 0x88000>;
+            clocks = <&adma_lpcg 0>, <&adma_lpcg 1>, <&adma_lpcg 2>;
+            clock-names = "ipg", "ocram", "core";
+            power-domains = <&pd 0>, <&pd 1>, <&pd 2>, <&pd 3>;
+            mbox-names = "txdb0", "txdb1", "rxdb0", "rxdb1";
+            mboxes = <&mhu_tx 0 2 0>, //TX - DBE, Channel Window 2, Doorbel 0
+                     <&mhu_tx 0 2 1>, //TX - DBE, Chanel Window 2, Doorbell 1
+                     <&mhu_rx 0 2 0>, //RX - DBE, Channel Window 2, Doorbell 0
+                     <&mhu_rx 0 2 1>; //RX - DBE, Channel Window 2, Doorbell 1
+            memory-region = <&dsp_reserved>;
+        };
+    };
-- 
2.34.1

